airflow/models/mappedoperator.py

```diff
@@ -578,10 +584,14 @@ def expand_mapped_task(self, run_id: str, *, session: Session) -> Sequence["Task
         return ret
 
     def prepare_for_execution(self) -> "MappedOperator":
-        # Since a mapped operator cannot be used for execution, and an unmapped
-        # BaseOperator needs to be created later (see render_template_fields),
-        # we don't need to create a copy of the MappedOperator here.
-        return self
+        # Ensure that we return a copy, so that in any executor where the
+        # mapped task is expanded in the same process (just DebugExecutor right now) we don't mutate the
+        # MappedOperator in the `unmap` function affecting subsequent mapped TIs of this task
+        other = copy.copy(self)
+        assert other.task_group
```

Comment:
Maybe a note saying this is only called when a task is executed, and the task must be in a DAG at that point. (Or perhaps we can raise `if task_group is None`.)